core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java

```diff
@@ -0,0 +1,265 @@
+package org.wso2.carbon.user.core.util;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.tomcat.jdbc.pool.PoolProperties;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.wso2.carbon.user.api.RealmConfiguration;
+import org.wso2.carbon.user.api.UserStoreException;
+
+import javax.sql.DataSource;
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.*;
+
+public class DatabaseUtilTest {
+
+    private static Log log = LogFactory.getLog(DatabaseUtil.class);
+
+    @Mock DataSource datasource;
+    @Mock Connection conn;
+    @Mock PreparedStatement preparedStatement;
+    @Mock Statement statement;
+    @Mock ResultSet resultSetMock;
+    @Mock RealmConfiguration realmConfig;
+    @Mock PoolProperties poolProperties;
+    @Mock CallableStatement callableStatementMock;
+
+    @Before
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
+        when(datasource.getConnection()).thenReturn(conn);
+        when(conn.prepareStatement(any(String.class))).thenReturn(preparedStatement);
+        when(statement.getResultSet()).thenReturn(resultSetMock);
+    }
+
+    @Test
+    public void getDBConnectionShouldNotBeNull() throws Exception {
+        // not a true unit test.
+        String url = "jdbc:h2:target/ReadOnlyTest/CARBON_TEST";
+        String driverName = "org.h2.Driver";
+        PoolProperties poolProperties = new PoolProperties();
+        poolProperties.setUrl(url);
+        poolProperties.setDriverClassName(driverName);
+        DataSource fake = new org.apache.tomcat.jdbc.pool.DataSource(poolProperties);
+
+        Assert.assertNotNull(DatabaseUtil.getDBConnection(fake));
+    }
+
+    @Test
+    public void closeConnectionShouldNotBeNull() throws SQLException {
+
+        Mockito.doAnswer(RETURNS_MOCKS).when(datasource).getConnection();
+        DatabaseUtil.closeConnection(conn);
+        assertNotNull(conn);
+    }
+
+    @Test
+    public void closeConnectionShouldHandlesSQLRecoverableException() throws Exception {
+
+            Mockito.doThrow(new SQLRecoverableException("throw SQLRecoverableException")).when(conn).close();
+            DatabaseUtil.closeConnection(conn);
+            System.out.println("Expected error thrown in sqlerecoverable handler");
+            Assert.assertNotNull(conn);

```

Comment:
Would you mind explaining how this assertion works? AFAIU, database connection object would never be null irrespective of the actual database connection is active or not. Thanks